lang/ghc/Portfile

```diff
@@ -29,144 +24,173 @@ long_description    \
-set llvmPrefix  ${prefix}/libexec/llvm-3.5
-configure.args  --with-ghc=${bootstraproot}/bin/ghc \
-                --with-iconv-includes=${prefix}/include \
-                --with-iconv-libraries=${prefix}/lib \
-                --with-gmp-includes=${prefix}/include \
-                --with-gmp-libraries=${prefix}/lib \
-                --with-ffi-includes=${prefix}/lib/libffi-3.2.1/include \
-                --with-ffi-libraries=${prefix}/lib \
-                --with-system-libffi \
-                --with-gcc="${configure.cc}" \
-                --with-llc=${llvmPrefix}/bin/llc \
-                --with-opt=${llvmPrefix}/bin/opt
-
-# OK so because the bootstrap binary has been prebuilt for libraries in
-# /usr/lib we search these in addition to MacPorts' paths to prevent link
-# errors. GHC _should_ actually compile itself in stage2 using paths from the
-# command line arguments
-compiler.cpath /usr/include
-compiler.library_path /usr/lib
-
-build.args      VERBOSE=1
-destroot.args   VERBOSE=1
-
-pre-activate {
-    set obsoletes [list]
-
-    # Legacy port deactivation hack added 2012-12-08, hs-platform-ghc no longer exists
-    lappend obsoletes hs-platform-ghc
-    # Legacy port deactivation hack added 2013-08-11, hs-process is provided by base
-    lappend obsoletes hs-process
-    # Legacy port deactivation hack added 2013-08-14, hs-template-haskell is provided by base
-    lappend obsoletes hs-template-haskell
-    # Legacy port deactivation hack added 2015-01-05, hs-transformers, hs-xhtml
-    # and hs-binary are provided by base
-    lappend obsoletes hs-transformers
-    lappend obsoletes hs-xhtml
-    lappend obsoletes hs-binary
-
-    foreach obsolete $obsoletes {
-        if {![catch {set obsolete_installed [lindex [registry_active $obsolete] 0]}]} {
-            # $obsolete is installed and active
-            # force deactivation
-            registry_deactivate_composite $obsolete "" [list ports_nodepcheck 1]
-        }
+destroot {
+    ui_msg "\tBuild/destroot from ${name}-${version}-x86_64-apple-darwin${extract.suffix}…"
+    system -W ${worksrcpath} "make install -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
+
+    if {[variant_isset "bootstrap"]} {
+        ui_msg "\tBootstrap build/destroot from ${name}-${version}-src${extract.suffix}…"
+        xinstall -m 0755 -d ${workpath}/src
+        system -W ${workpath}/src \
+            "tar xJf ${distpath}/${name}-${version}-src${extract.suffix}"
+        system -W ${workpath}/src/${name}-${version} \
+            "./boot"
+        system -W ${workpath}/src/${name}-${version} \
+            "export PATH=\${PATH}:${destroot}${prefix}/bin ; \
+                ./configure --prefix=${destroot}${prefix}"
+        system -W ${workpath}/src/${name}-${version} \
+            "make -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
+        system -W ${workpath}/src/${name}-${version} \
+            "make install -j\$((`/usr/bin/sysctl -n hw.physicalcpu`+1))"
```

Comment:
Both steps would be automatically executed for you, including the correct number of parallel jobs, if you would use a subport for the purpose.

Reply:
Thanks, but I still don’t know enough to implement a subport approach.

What’s a subport?

How does it work, exactly?

Right now, I’m using the binary provided in the first stage, saved in ${workpath}, to build the second stage, then install the second stage after it’s done.

With a separate port, I’ll have to worry about destrooting the 1st stage, installing over it, communicating between two ports via the file system, having two ports control the same file, and other stuff.

That sounds a **lot** more complicated than just running a few build command in an if block, as currently proposed.